[Bug]: TreeNode has already been completed when open upsert mode and use insert into in Spark

### What happened?

I am doing a performance test of spark upsert mode in the Arctic.
I have a table like this:
```
create table lineitem_upsert_41 (
   l_orderkey bigint,
   l_partkey bigint,
   l_suppkey bigint,
   l_linenumber int,
   l_quantity double,
   l_extendedprice double,
   l_discount double,
   l_tax double,
   l_returnflag string,
   l_linestatus string,
   l_shipdate date,
   l_receiptdate date,
   l_shipinstruct string,
   l_shipmode string,
   l_comment string,
   l_commitdate date,
   PRIMARY KEY (l_orderkey,l_linenumber)
) USING arctic PARTITIONED BY (l_commitdate) tblproperties (
	"write.upsert.enabled"="true",
  "base.file-index.hash-bucket"="4"
);
```
This is a TPCH table, and I set these session properties:
```
set `spark.sql.arctic.check-source-data-uniqueness.enabled`=`true`
set `spark.sql.arctic.optimize-write-enabled`=`true`
```

First, I insert overwrite some data(sf100 in TPCH, about 15GB) into the table.The SQL is :
```
insert overwrite table lineitem_upsert_41 select nvl(l_orderkey,0),l_partkey,l_suppkey,nvl(l_linenumber,0),l_quantity,l_extendedprice,l_discount,l_tax,l_returnflag,l_linestatus,l_shipdate,l_receiptdate,l_shipinstruct,l_shipmode,l_comment,l_commitdate from spark_catalog.test_origin.lineitem_sf100;
```
Insert overwrite works fine. And when I insert into some data(sf1, about 150 MB) the table use SQL:
```
insert into table lineitem_upsert_41 select nvl(l_orderkey,0),l_partkey,l_suppkey,nvl(l_linenumber,0),l_quantity,l_extendedprice,l_discount,l_tax,l_returnflag,l_linestatus,l_shipdate,l_receiptdate,l_shipinstruct,l_shipmode,l_comment,l_commitdate from spark_catalog.test_origin.lineitem_sf1;
```
I will get the error.
```
Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2303)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2252)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2251)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2251)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1124)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1124)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1124)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2490)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2432)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2421)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:902)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2196)
	at com.netease.arctic.shade.org.apache.spark.sql.execution.datasources.v2.ArcticTableWriteExec.writeInsert(ArcticTableWriteExec.scala:78)
	... 31 more
Caused by: java.lang.IllegalStateException: The write key [[8075] TreeNode(3,3)] has already been completed
	at com.netease.arctic.io.writer.BaseTaskWriter$OrderedWriterHolder.getDataWriter(BaseTaskWriter.java:292)
	at com.netease.arctic.io.writer.BaseTaskWriter$WriterHolder.get(BaseTaskWriter.java:187)
	at com.netease.arctic.io.writer.BaseTaskWriter.write(BaseTaskWriter.java:81)
	at com.netease.arctic.spark.writer.SimpleKeyedUpsertDataWriter.write(SimpleKeyedUpsertDataWriter.java:38)
	at com.netease.arctic.spark.writer.SimpleKeyedUpsertDataWriter.write(SimpleKeyedUpsertDataWriter.java:19)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$1(WriteToDataSourceV2Exec.scala:416)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:452)
	at com.netease.arctic.shade.org.apache.spark.sql.execution.datasources.v2.ArcticTableWriteExec.$anonfun$writeInsert$2(ArcticTableWriteExec.scala:81)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:498)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:501)
	... 3 more

	at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:69)
	at org.apache.kyuubi.operation.ExecuteStatement.waitStatementComplete(ExecuteStatement.scala:119)
	at org.apache.kyuubi.operation.ExecuteStatement.$anonfun$runInternal$1(ExecuteStatement.scala:154)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748) (state=,code=0)
```

### Affects Versions

master

### What engines are you seeing the problem on?

Spark

### How to reproduce

According to 'What happened?'.Just execute the SQL, and the source table is the TPCH's `lineitem`.

### Relevant log output

```shell
Driver stacktrace:
	at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2303)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2252)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2251)
	at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
	at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
	at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2251)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1124)
	at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1124)
	at scala.Option.foreach(Option.scala:407)
	at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1124)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:2490)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2432)
	at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2421)
	at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
	at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:902)
	at org.apache.spark.SparkContext.runJob(SparkContext.scala:2196)
	at com.netease.arctic.shade.org.apache.spark.sql.execution.datasources.v2.ArcticTableWriteExec.writeInsert(ArcticTableWriteExec.scala:78)
	... 31 more
Caused by: java.lang.IllegalStateException: The write key [[8075] TreeNode(3,3)] has already been completed
	at com.netease.arctic.io.writer.BaseTaskWriter$OrderedWriterHolder.getDataWriter(BaseTaskWriter.java:292)
	at com.netease.arctic.io.writer.BaseTaskWriter$WriterHolder.get(BaseTaskWriter.java:187)
	at com.netease.arctic.io.writer.BaseTaskWriter.write(BaseTaskWriter.java:81)
	at com.netease.arctic.spark.writer.SimpleKeyedUpsertDataWriter.write(SimpleKeyedUpsertDataWriter.java:38)
	at com.netease.arctic.spark.writer.SimpleKeyedUpsertDataWriter.write(SimpleKeyedUpsertDataWriter.java:19)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.$anonfun$run$1(WriteToDataSourceV2Exec.scala:416)
	at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
	at org.apache.spark.sql.execution.datasources.v2.DataWritingSparkTask$.run(WriteToDataSourceV2Exec.scala:452)
	at com.netease.arctic.shade.org.apache.spark.sql.execution.datasources.v2.ArcticTableWriteExec.$anonfun$writeInsert$2(ArcticTableWriteExec.scala:81)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:131)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:498)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:501)
	... 3 more

	at org.apache.kyuubi.KyuubiSQLException$.apply(KyuubiSQLException.scala:69)
	at org.apache.kyuubi.operation.ExecuteStatement.waitStatementComplete(ExecuteStatement.scala:119)
	at org.apache.kyuubi.operation.ExecuteStatement.$anonfun$runInternal$1(ExecuteStatement.scala:154)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748) (state=,code=0)
```


### Anything else

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/spark/v3.1/spark/src/main/java/com/netease/arctic/spark/util/ArcticSparkUtils.java
+++ b/spark/v3.1/spark/src/main/java/com/netease/arctic/spark/util/ArcticSparkUtils.java
@@ -18,10 +18,8 @@
 
 package com.netease.arctic.spark.util;
 
-import com.netease.arctic.table.ArcticTable;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.util.Utf8;
-import org.apache.iceberg.PartitionSpec;
 import org.apache.iceberg.relocated.com.google.common.base.Joiner;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.spark.Spark3Util;
@@ -38,9 +36,7 @@ import org.slf4j.LoggerFactory;
 
 import java.math.BigDecimal;
 import java.nio.ByteBuffer;
-import java.util.HashSet;
 import java.util.List;
-import java.util.stream.Collectors;
 
 public class ArcticSparkUtils {
   private static final Logger LOG = LoggerFactory.getLogger(ArcticSparkUtils.class);
@@ -99,21 +95,5 @@ public class ArcticSparkUtils {
       default:
     }
     return value;
-  }
-
-  public static boolean isPrimaryKeyIncludeAllPartitions(ArcticTable table) {
-    Preconditions.checkArgument(table.isKeyedTable(), "it's must be a keyed table");
-    List<String> pkCols = table.asKeyedTable().primaryKeySpec().fieldNames();
-    PartitionSpec ptSpec = table.spec();
-
-    boolean withTransform = ptSpec.fields().stream()
-        .anyMatch(f -> !f.transform().isIdentity());
-    if (withTransform) {
-      return false;
-    }
-    List<String> ptCols = ptSpec.fields().stream()
-        .map(f -> table.schema().findField(f.sourceId()).name())
-        .collect(Collectors.toList());
-    return new HashSet<>(pkCols).containsAll(ptCols);
   }
 }

--- a/spark/v3.1/spark/src/main/java/com/netease/arctic/spark/util/ArcticSparkUtils.java
+++ b/spark/v3.1/spark/src/main/java/com/netease/arctic/spark/util/ArcticSparkUtils.java
@@ -18,27 +18,19 @@
 
 package com.netease.arctic.spark.util;
 
-import com.netease.arctic.spark.table.ArcticSparkTable;
-import com.netease.arctic.table.DistributionHashMode;
-import com.netease.arctic.table.PrimaryKeySpec;
-import com.netease.arctic.table.TableProperties;
+import com.netease.arctic.table.ArcticTable;
 import org.apache.avro.generic.GenericData;
 import org.apache.avro.util.Utf8;
-import org.apache.iceberg.DistributionMode;
+import org.apache.iceberg.PartitionSpec;
 import org.apache.iceberg.relocated.com.google.common.base.Joiner;
 import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
 import org.apache.iceberg.spark.Spark3Util;
 import org.apache.iceberg.types.Type;
 import org.apache.iceberg.util.ByteBuffers;
-import org.apache.iceberg.util.PropertyUtil;
 import org.apache.spark.sql.SparkSession;
 import org.apache.spark.sql.connector.catalog.CatalogPlugin;
 import org.apache.spark.sql.connector.catalog.Identifier;
 import org.apache.spark.sql.connector.catalog.TableCatalog;
-import org.apache.spark.sql.connector.expressions.Expressions;
-import org.apache.spark.sql.connector.expressions.Transform;
-import org.apache.spark.sql.connector.iceberg.distributions.Distribution;
-import org.apache.spark.sql.connector.iceberg.distributions.Distributions;
 import org.apache.spark.sql.types.Decimal;
 import org.apache.spark.unsafe.types.UTF8String;
 import org.slf4j.Logger;
@@ -46,14 +38,9 @@ import org.slf4j.LoggerFactory;
 
 import java.math.BigDecimal;
 import java.nio.ByteBuffer;
-import java.util.ArrayList;
-import java.util.Arrays;
+import java.util.HashSet;
 import java.util.List;
-import java.util.Objects;
-
-import static com.netease.arctic.table.TableProperties.WRITE_DISTRIBUTION_MODE;
-import static com.netease.arctic.table.TableProperties.WRITE_DISTRIBUTION_MODE_DEFAULT;
-import static org.apache.iceberg.spark.Spark3Util.toTransforms;
+import java.util.stream.Collectors;
 
 public class ArcticSparkUtils {
   private static final Logger LOG = LoggerFactory.getLogger(ArcticSparkUtils.class);
@@ -86,61 +73,6 @@ public class ArcticSparkUtils {
     }
   }
 
-  public static Distribution buildRequiredDistribution(ArcticSparkTable arcticSparkTable) {
-    // Fallback to use distribution mode parsed from table properties .
-    String modeName = PropertyUtil.propertyAsString(
-        arcticSparkTable.properties(),
-        WRITE_DISTRIBUTION_MODE,
-        WRITE_DISTRIBUTION_MODE_DEFAULT);
-    DistributionMode writeMode = DistributionMode.fromName(modeName);
-    switch (writeMode) {
-      case NONE:
-        return null;
-
-      case HASH:
-        DistributionHashMode distributionHashMode = DistributionHashMode.valueOfDesc(
-            arcticSparkTable.properties().getOrDefault(
-                TableProperties.WRITE_DISTRIBUTION_HASH_MODE,
-                TableProperties.WRITE_DISTRIBUTION_HASH_MODE_DEFAULT));
-        List<Transform> transforms = new ArrayList<>();
-        if (DistributionHashMode.AUTO.equals(distributionHashMode)) {
-          distributionHashMode = DistributionHashMode.autoSelect(
-              arcticSparkTable.table().isKeyedTable(),
-              !arcticSparkTable.table().spec().isUnpartitioned());
-        }
-        if (distributionHashMode.isSupportPrimaryKey()) {
-          Transform transform = toTransformsFromPrimary(
-              arcticSparkTable,
-              arcticSparkTable.table().asKeyedTable().primaryKeySpec());
-          transforms.add(transform);
-          if (distributionHashMode.isSupportPartition()) {
-            transforms.addAll(Arrays.asList(toTransforms(arcticSparkTable.table().spec())));
-          }
-          return Distributions.clustered(transforms.stream().filter(Objects::nonNull).toArray(Transform[]::new));
-        } else {
-          if (distributionHashMode.isSupportPartition()) {
-            return Distributions.clustered(toTransforms(arcticSparkTable.table().spec()));
-          } else {
-            return null;
-          }
-        }
-
-      case RANGE:
-        LOG.warn("Fallback to use 'none' distribution mode, because {}={} is not supported in spark now",
-            WRITE_DISTRIBUTION_MODE, DistributionMode.RANGE.modeName());
-        return null;
-
-      default:
-        throw new RuntimeException("Unrecognized write.distribution-mode: " + writeMode);
-    }
-  }
-
-  private static Transform toTransformsFromPrimary(ArcticSparkTable arcticSparkTable, PrimaryKeySpec primaryKeySpec) {
-    int numBucket = PropertyUtil.propertyAsInt(arcticSparkTable.properties(),
-        TableProperties.BASE_FILE_INDEX_HASH_BUCKET, TableProperties.BASE_FILE_INDEX_HASH_BUCKET_DEFAULT);
-    return Expressions.bucket(numBucket, primaryKeySpec.fieldNames().get(0));
-  }
-
   public static Object convertConstant(Type type, Object value) {
     if (value == null) {
       return null;
@@ -167,5 +99,21 @@ public class ArcticSparkUtils {
       default:
     }
     return value;
+  }
+
+  public static boolean isPrimaryKeyIncludeAllPartitions(ArcticTable table) {
+    Preconditions.checkArgument(table.isKeyedTable(), "it's must be a keyed table");
+    List<String> pkCols = table.asKeyedTable().primaryKeySpec().fieldNames();
+    PartitionSpec ptSpec = table.spec();
+
+    boolean withTransform = ptSpec.fields().stream()
+        .anyMatch(f -> !f.transform().isIdentity());
+    if (withTransform) {
+      return false;
+    }
+    List<String> ptCols = ptSpec.fields().stream()
+        .map(f -> table.schema().findField(f.sourceId()).name())
+        .collect(Collectors.toList());
+    return new HashSet<>(pkCols).containsAll(ptCols);
   }
 }

--- a/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
+++ b/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
@@ -19,9 +19,9 @@
 package com.netease.arctic.spark.sql.catalyst.optimize
 
 import com.netease.arctic.spark.sql.ArcticExtensionUtils.{isArcticIcebergRelation, isArcticRelation}
-import com.netease.arctic.spark.sql.catalyst.plans.{AppendArcticData, OverwriteArcticPartitionsDynamic, ReplaceArcticData}
+import com.netease.arctic.spark.sql.catalyst.plans.OverwriteArcticPartitionsDynamic
 import com.netease.arctic.spark.table.{ArcticIcebergSparkTable, ArcticSparkTable}
-import com.netease.arctic.spark.util.{ArcticSparkUtils, DistributionAndOrderingUtil}
+import com.netease.arctic.spark.util.DistributionAndOrderingUtil
 import com.netease.arctic.spark.{SparkSQLProperties, SupportSparkAdapter}
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.expressions.{Expression, SortOrder}
@@ -61,18 +61,6 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
       val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
       a.copy(query = newQuery, writeOptions = options)
 
-    case a @ AppendArcticData(r: DataSourceV2Relation, query, _, writeOptions)
-      if isArcticRelation(r) && optimizeForRowLevelOperation(r) =>
-      val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
-      val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
-      a.copy(query = newQuery, writeOptions = options)
-
-    case a @ ReplaceArcticData(r: DataSourceV2Relation, query, writeOptions)
-      if isArcticRelation(r) && optimizeForRowLevelOperation(r) =>
-      val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
-      val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
-      a.copy(query = newQuery, writeOptions = options)
-
     case o @ OverwriteArcticPartitionsDynamic(r: DataSourceV2Relation, query, _, writeOptions)
       if isArcticRelation(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false)
@@ -93,16 +81,6 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
       if isArcticIcebergRelation(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false)
       o.copy(query = newQuery)
-  }
-
-  import com.netease.arctic.spark.sql.ArcticExtensionUtils._
-
-  def optimizeForRowLevelOperation(r: DataSourceV2Relation): Boolean = {
-    val table = r.table.asArcticTable
-    if (table.table().isUnkeyedTable) {
-      return false
-    }
-    ArcticSparkUtils.isPrimaryKeyIncludeAllPartitions(table.table())
   }
 
   def optimizeWriteEnabled(): Boolean = {

--- a/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
+++ b/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
@@ -62,13 +62,13 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
       a.copy(query = newQuery, writeOptions = options)
 
     case a @ AppendArcticData(r: DataSourceV2Relation, query, _, writeOptions)
-      if isArcticRelation(r) && isPrimaryKeyIncludePartitions(r) =>
+      if isArcticRelation(r) && optimizeForRowLevelOperation(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
       val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
       a.copy(query = newQuery, writeOptions = options)
 
     case a @ ReplaceArcticData(r: DataSourceV2Relation, query, writeOptions)
-      if isArcticRelation(r) && isPrimaryKeyIncludePartitions(r) =>
+      if isArcticRelation(r) && optimizeForRowLevelOperation(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
       val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
       a.copy(query = newQuery, writeOptions = options)
@@ -97,8 +97,11 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
 
   import com.netease.arctic.spark.sql.ArcticExtensionUtils._
 
-  def isPrimaryKeyIncludePartitions(r: DataSourceV2Relation): Boolean = {
+  def optimizeForRowLevelOperation(r: DataSourceV2Relation): Boolean = {
     val table = r.table.asArcticTable
+    if (table.table().isUnkeyedTable) {
+      return false
+    }
     ArcticSparkUtils.isPrimaryKeyIncludeAllPartitions(table.table())
   }
 

--- a/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
+++ b/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/OptimizeWriteRule.scala
@@ -19,9 +19,9 @@
 package com.netease.arctic.spark.sql.catalyst.optimize
 
 import com.netease.arctic.spark.sql.ArcticExtensionUtils.{isArcticIcebergRelation, isArcticRelation}
-import com.netease.arctic.spark.sql.catalyst.plans.{AppendArcticData, OverwriteArcticPartitionsDynamic}
+import com.netease.arctic.spark.sql.catalyst.plans.{AppendArcticData, OverwriteArcticPartitionsDynamic, ReplaceArcticData}
 import com.netease.arctic.spark.table.{ArcticIcebergSparkTable, ArcticSparkTable}
-import com.netease.arctic.spark.util.DistributionAndOrderingUtil
+import com.netease.arctic.spark.util.{ArcticSparkUtils, DistributionAndOrderingUtil}
 import com.netease.arctic.spark.{SparkSQLProperties, SupportSparkAdapter}
 import org.apache.spark.sql.SparkSession
 import org.apache.spark.sql.catalyst.expressions.{Expression, SortOrder}
@@ -62,7 +62,13 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
       a.copy(query = newQuery, writeOptions = options)
 
     case a @ AppendArcticData(r: DataSourceV2Relation, query, _, writeOptions)
-      if isArcticRelation(r) =>
+      if isArcticRelation(r) && isPrimaryKeyIncludePartitions(r) =>
+      val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
+      val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
+      a.copy(query = newQuery, writeOptions = options)
+
+    case a @ ReplaceArcticData(r: DataSourceV2Relation, query, writeOptions)
+      if isArcticRelation(r) && isPrimaryKeyIncludePartitions(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false, writeBase = false)
       val options = writeOptions + ("writer.distributed-and-ordered" -> "true")
       a.copy(query = newQuery, writeOptions = options)
@@ -87,6 +93,13 @@ case class OptimizeWriteRule(spark: SparkSession) extends Rule[LogicalPlan] with
       if isArcticIcebergRelation(r) =>
       val newQuery = distributionQuery(query, r.table, rowLevelOperation = false)
       o.copy(query = newQuery)
+  }
+
+  import com.netease.arctic.spark.sql.ArcticExtensionUtils._
+
+  def isPrimaryKeyIncludePartitions(r: DataSourceV2Relation): Boolean = {
+    val table = r.table.asArcticTable
+    ArcticSparkUtils.isPrimaryKeyIncludeAllPartitions(table.table())
   }
 
   def optimizeWriteEnabled(): Boolean = {

--- a/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/RewriteAppendArcticTable.scala
+++ b/spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/catalyst/optimize/RewriteAppendArcticTable.scala
@@ -141,14 +141,14 @@ case class RewriteAppendArcticTable(spark: SparkSession) extends Rule[LogicalPla
         SparkSQLProperties.CHECK_SOURCE_DUPLICATES_ENABLE_DEFAULT))
   }
 
-  def buildJoinCondition(primaries: util.List[String], r: DataSourceV2Relation, insertPlan: LogicalPlan): Expression = {
+  def buildJoinCondition(primaries: util.List[String], tableScan: LogicalPlan, insertPlan: LogicalPlan): Expression = {
     var i = 0
     var joinCondition: Expression = null
     val expressions = new util.ArrayList[Expression]
     while (i < primaries.size) {
       val primary = primaries.get(i)
-      val primaryAttr = r.output.find(_.name == primary).get
-      val joinAttribute = insertPlan.output.find(_.name.replace("_arctic_after_", "") == primary).get
+      val primaryAttr = insertPlan.output.find(_.name == primary).get
+      val joinAttribute = tableScan.output.find(_.name.replace("_arctic_before_", "") == primary).get
       val experssion = EqualTo(primaryAttr, joinAttribute)
       expressions.add(experssion)
       i += 1
@@ -171,9 +171,10 @@ case class RewriteAppendArcticTable(spark: SparkSession) extends Rule[LogicalPla
       case arctic: ArcticSparkTable =>
         if (arctic.table().isKeyedTable) {
           val primaries = arctic.table().asKeyedTable().primaryKeySpec().fieldNames()
-          val insertPlan = buildKeyedTableInsertProjection(query)
-          val joinCondition = buildJoinCondition(primaries, r, insertPlan)
-          Join(r, insertPlan, RightOuter, Some(joinCondition), JoinHint.NONE)
+          val tablePlan = buildKeyedTableBeforeProject(r)
+          // val insertPlan = buildKeyedTableInsertProjection(query)
+          val joinCondition = buildJoinCondition(primaries, tablePlan, query)
+          Join(tablePlan, query, RightOuter, Some(joinCondition), JoinHint.NONE)
         } else {
           query
         }
@@ -187,4 +188,13 @@ case class RewriteAppendArcticTable(spark: SparkSession) extends Rule[LogicalPla
     })
     Project(outputWithValues, relation)
   }
+
+  private def buildKeyedTableBeforeProject(relation: DataSourceV2Relation): LogicalPlan = {
+    val output = relation.output
+    val outputWithValues = output.map(a => {
+      Alias(a, "_arctic_before_" + a.name)()
+    })
+    Project(outputWithValues, relation)
+  }
+
 }

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/ArcticSparkCatalogTestGroup.java
@@ -77,7 +77,9 @@ import java.util.Map;
     TestHiveTableDropPartitions.class,
     TestHiveTableTruncate.class,
     TestKeyedTableMergeInto.class,
-    TestUnKeyedTableMergeInto.class})
+    TestUnKeyedTableMergeInto.class,
+    TestUpsert.class
+})
 public class ArcticSparkCatalogTestGroup {
 
   @BeforeClass

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/SparkTestContext.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/SparkTestContext.java
@@ -625,4 +625,12 @@ public class SparkTestContext extends ExternalResource {
     }
     return RowFactory.create(values);
   }
+
+  public Map<String, String> properties(String... kv) {
+    Map<String, String> props = Maps.newHashMap();
+    for (int i = 0; i < kv.length; i = i + 2) {
+      props.put(kv[i], kv[i + 1]);
+    }
+    return props;
+  }
 }

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
@@ -18,7 +18,6 @@
 
 package com.netease.arctic.spark;
 
-import com.clearspring.analytics.util.Lists;
 import com.netease.arctic.catalog.ArcticCatalog;
 import com.netease.arctic.spark.util.RecordGenerator;
 import com.netease.arctic.table.PrimaryKeySpec;
@@ -26,94 +25,83 @@ import com.netease.arctic.table.TableIdentifier;
 import org.apache.iceberg.PartitionSpec;
 import org.apache.iceberg.Schema;
 import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
 import org.apache.iceberg.spark.SparkSchemaUtil;
 import org.apache.iceberg.types.Types;
 import org.apache.spark.sql.AnalysisException;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Row;
 import org.glassfish.jersey.internal.guava.Sets;
+import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
 
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
 
+@RunWith(Parameterized.class)
 public class TestUpsert extends SparkTestBase {
   private final String database = "db";
   private final String table = "sink_table";
-  private final String sourceTable = "source_table";
+  private final String sourceView = "source_table";
   private final String initView = "init_view";
   private final TableIdentifier identifier = TableIdentifier.of(catalogNameArctic, database, table);
 
-  private List<GenericRecord> sources;
-  private List<GenericRecord> initialize;
+  private static final Schema schema = new Schema(
+      Types.NestedField.of(1, false, "order_key", Types.IntegerType.get()),
+      Types.NestedField.of(2, false, "line_number", Types.IntegerType.get()),
+      Types.NestedField.of(3, false, "data", Types.StringType.get()),
+      Types.NestedField.of(4, false, "pt", Types.StringType.get())
+  );
 
-  private Schema schema;
-  private int newRecordSize = 300;
-  private int upsertRecordSize = 200;
+  private static final PartitionSpec partitionSpec = PartitionSpec.builderFor(schema)
+      .identity("pt").build();
 
-  @Before
-  public void before() throws AnalysisException {
-    sql("use " + catalogNameArctic);
-    sql("create database if not exists {0}", database);
+  private static final int newRecordSize = 300;
+  private static final int upsertRecordSize = 200;
 
-    schema = new Schema(
-        Types.NestedField.of(1, false, "order_key", Types.IntegerType.get()),
-        Types.NestedField.of(2, false, "line_number", Types.IntegerType.get()),
-        Types.NestedField.of(3, false, "data", Types.StringType.get()),
-        Types.NestedField.of(4, false, "pt", Types.StringType.get())
+  @Parameterized.Parameters
+  public static List<Object[]> arguments() {
+    // pk-include-pt, duplicateCheck
+    List<Boolean[]> args = Lists.newArrayList(
+        new Boolean[] {true, true},
+        new Boolean[] {false, true},
+        new Boolean[] {true, false},
+        new Boolean[] {false, false}
     );
 
-    ArcticCatalog catalog = catalog(catalogNameArctic);
-    PrimaryKeySpec primaryKeySpec = PrimaryKeySpec.builderFor(schema)
-        .addColumn("order_key")
-        .addColumn("line_number").build();
+    return args.stream()
+        .map(a -> {
+          boolean pkIncludePt = a[0];
+          boolean duplicateCheck = a[1];
+          PrimaryKeySpec primaryKeySpec = PrimaryKeySpec.builderFor(schema)
+              .addColumn("order_key")
+              .addColumn("line_number").build();
+          if (pkIncludePt) {
+            primaryKeySpec = PrimaryKeySpec.builderFor(schema)
+                .addColumn("order_key")
+                .addColumn("line_number")
+                .addColumn("pt").build();
+          }
+          RecordGenerator generator = RecordGenerator.buildFor(schema)
+              .withSequencePrimaryKey(primaryKeySpec)
+              .withRandomDate("pt")
+              .build();
+          List<GenericRecord> target = generator.records(1000);
+          List<GenericRecord> source = upsertSource(
+              target, generator, primaryKeySpec, partitionSpec, newRecordSize, upsertRecordSize);
 
-    PartitionSpec partitionSpec = PartitionSpec.builderFor(schema)
-        .identity("pt").build();
-    catalog.newTableBuilder(identifier, schema)
-        .withPrimaryKeySpec(primaryKeySpec)
-        .withPartitionSpec(partitionSpec)
-        .withProperties(properties(
-            "write.upsert.enabled", "true",
-            "base.file-index.hash-bucket", "4"
-        )).create();
-
-    RecordGenerator generator = RecordGenerator.buildFor(schema)
-        .withSequencePrimaryKey(primaryKeySpec)
-        .withRandomDate("pt")
-        .build();
-    initialize = generator.records(1000);
-
-    Dataset<Row> dataset = spark.createDataFrame(
-        initialize.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
-        SparkSchemaUtil.convert(schema));
-    dataset.createTempView(initView);
-
-    sources = upsertSource(initialize, generator,
-        primaryKeySpec, partitionSpec, upsertRecordSize, newRecordSize);
-    dataset = spark.createDataFrame(
-        sources.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
-        SparkSchemaUtil.convert(schema));
-    dataset.createTempView(sourceTable);
+          return new Object[] {target, source, primaryKeySpec, duplicateCheck};
+        })
+        .collect(Collectors.toList());
   }
 
-  @Test
-  public void testKeyedTableUpsert() {
-    sql("set `spark.sql.arctic.check-source-data-uniqueness.enabled`=`true`");
-    sql("set `spark.sql.arctic.optimize-write-enabled`=`true`");
-
-    sql("INSERT OVERWRITE " + database + "." + table + " SELECT * FROM " + initView);
-    sql("insert into table " + database + '.' + table + " SELECT * FROM " + sourceTable);
-
-    rows = sql("SELECT * FROM " + database + "." + table);
-    Assert.assertEquals(initialize.size() + newRecordSize, rows.size());
-  }
-
-  private List<GenericRecord> upsertSource(
+  private static List<GenericRecord> upsertSource(
       List<GenericRecord> initializeData, RecordGenerator sourceGenerator,
       PrimaryKeySpec keySpec, PartitionSpec partitionSpec,
       int upsertCount, int insertCount) {
@@ -130,15 +118,77 @@ public class TestUpsert extends SparkTestBase {
 
     Set<String> sourceCols = Sets.newHashSet();
     sourceCols.addAll(keySpec.fieldNames());
-    // partitionSpec.fields().stream()
-    //     .map(f -> schema.findField(f.sourceId()))
-    //     .map(Types.NestedField::name)
-    //     .forEach(sourceCols::add);
+    partitionSpec.fields().stream()
+        .map(f -> schema.findField(f.sourceId()))
+        .map(Types.NestedField::name)
+        .forEach(sourceCols::add);
 
     for (GenericRecord r : upsertSource) {
       GenericRecord t = it.next();
       sourceCols.forEach(col -> t.setField(col, r.getField(col)));
     }
     return source;
+  }
+
+  private final List<GenericRecord> sources;
+  private final List<GenericRecord> initialize;
+  private final boolean checkSourceUniqueness;
+  private final PrimaryKeySpec primaryKeySpec;
+
+  public TestUpsert(
+      List<GenericRecord> targetData, List<GenericRecord> sourceData,
+      PrimaryKeySpec primaryKeySpec,
+      boolean sourceUniquenessCheck
+  ) {
+    this.initialize = targetData;
+    this.sources = sourceData;
+    this.checkSourceUniqueness = sourceUniquenessCheck;
+    this.primaryKeySpec = primaryKeySpec;
+  }
+
+  @Before
+  public void before() throws AnalysisException {
+    sql("use " + catalogNameArctic);
+    sql("create database if not exists {0}", database);
+
+    ArcticCatalog catalog = catalog(catalogNameArctic);
+
+    catalog.newTableBuilder(identifier, schema)
+        .withPrimaryKeySpec(primaryKeySpec)
+        .withPartitionSpec(partitionSpec)
+        .withProperties(properties(
+            "write.upsert.enabled", "true",
+            "base.file-index.hash-bucket", "4"
+        )).create();
+
+    Dataset<Row> dataset = spark.createDataFrame(
+        initialize.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
+        SparkSchemaUtil.convert(schema));
+    dataset.createTempView(initView);
+
+    dataset = spark.createDataFrame(
+        sources.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
+        SparkSchemaUtil.convert(schema));
+    dataset.createTempView(sourceView);
+  }
+
+  @After
+  public void after() {
+    sql("DROP VIEW IF EXISTS " + initView);
+    sql("DROP VIEW IF EXISTS " + sourceView);
+    ArcticCatalog catalog = catalog(catalogNameArctic);
+    catalog.dropTable(identifier, true);
+  }
+
+  @Test
+  public void testKeyedTableUpsert() {
+    sql("set `spark.sql.arctic.check-source-data-uniqueness.enabled`=" + this.checkSourceUniqueness);
+    sql("set `spark.sql.arctic.optimize-write-enabled`=`true`");
+
+    sql("INSERT OVERWRITE " + database + "." + table + " SELECT * FROM " + initView);
+    sql("insert into table " + database + '.' + table + " SELECT * FROM " + sourceView);
+
+    rows = sql("SELECT * FROM " + database + "." + table);
+    Assert.assertEquals(initialize.size() + newRecordSize, rows.size());
   }
 }

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
@@ -1,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *  *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netease.arctic.spark;
+
+import com.clearspring.analytics.util.Lists;
+import com.netease.arctic.catalog.ArcticCatalog;
+import com.netease.arctic.spark.util.RecordGenerator;
+import com.netease.arctic.table.PrimaryKeySpec;
+import com.netease.arctic.table.TableIdentifier;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.spark.SparkSchemaUtil;
+import org.apache.iceberg.types.Types;
+import org.apache.spark.sql.AnalysisException;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.glassfish.jersey.internal.guava.Sets;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class TestUpsert extends SparkTestBase {
+  private final String database = "db";
+  private final String table = "sink_table";
+  private final String sourceTable = "source_table";
+  private final String initView = "init_view";
+  private final TableIdentifier identifier = TableIdentifier.of(catalogNameArctic, database, table);
+
+  private List<GenericRecord> sources;
+  private List<GenericRecord> initialize;
+
+  private Schema schema;
+  private int newRecordSize = 300;
+  private int upsertRecordSize = 200;
+
+  @Before
+  public void before() throws AnalysisException {
+    sql("use " + catalogNameArctic);
+    sql("create database if not exists {0}", database);
+
+    schema = new Schema(
+        Types.NestedField.of(1, false, "order_key", Types.IntegerType.get()),
+        Types.NestedField.of(2, false, "line_number", Types.IntegerType.get()),
+        Types.NestedField.of(3, false, "data", Types.StringType.get()),
+        Types.NestedField.of(4, false, "pt", Types.StringType.get())
+    );
+
+    ArcticCatalog catalog = catalog(catalogNameArctic);
+    PrimaryKeySpec primaryKeySpec = PrimaryKeySpec.builderFor(schema)
+        .addColumn("order_key")
+        .addColumn("line_number").build();
+
+    PartitionSpec partitionSpec = PartitionSpec.builderFor(schema)
+        .identity("pt").build();
+    catalog.newTableBuilder(identifier, schema)
+        .withPrimaryKeySpec(primaryKeySpec)
+        .withPartitionSpec(partitionSpec)
+        .withProperties(properties(
+            "write.upsert.enabled", "true",
+            "base.file-index.hash-bucket", "4"
+        )).create();
+
+    RecordGenerator generator = RecordGenerator.buildFor(schema)
+        .withSequencePrimaryKey(primaryKeySpec)
+        .withRandomDate("pt")
+        .build();
+    initialize = generator.records(1000);
+
+    Dataset<Row> dataset = spark.createDataFrame(
+        initialize.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
+        SparkSchemaUtil.convert(schema));
+    dataset.createTempView(initView);
+
+    sources = upsertSource(initialize, generator,
+        primaryKeySpec, partitionSpec, upsertRecordSize, newRecordSize);
+    dataset = spark.createDataFrame(
+        sources.stream().map(SparkTestContext::recordToRow).collect(Collectors.toList()),
+        SparkSchemaUtil.convert(schema));
+    dataset.createTempView(sourceTable);
+  }
+
+  @Test
+  public void testKeyedTableUpsert() {
+    sql("set `spark.sql.arctic.check-source-data-uniqueness.enabled`=`true`");
+    sql("set `spark.sql.arctic.optimize-write-enabled`=`true`");
+
+    sql("INSERT OVERWRITE " + database + "." + table + " SELECT * FROM " + initView);
+    sql("insert into table " + database + '.' + table + " SELECT * FROM " + sourceTable);
+
+    rows = sql("SELECT * FROM " + database + "." + table);
+    Assert.assertEquals(initialize.size() + newRecordSize, rows.size());
+  }
+
+  private List<GenericRecord> upsertSource(
+      List<GenericRecord> initializeData, RecordGenerator sourceGenerator,
+      PrimaryKeySpec keySpec, PartitionSpec partitionSpec,
+      int upsertCount, int insertCount) {
+    RecordGenerator generator = RecordGenerator.buildFor(schema)
+        .withSeed(System.currentTimeMillis())
+        .build();
+
+    List<GenericRecord> source = Lists.newArrayList();
+    List<GenericRecord> insertSource = sourceGenerator.records(insertCount);
+    source.addAll(insertSource);
+
+    List<GenericRecord> upsertSource = sourceGenerator.records(upsertCount);
+    Iterator<GenericRecord> it = initializeData.iterator();
+
+    Set<String> sourceCols = Sets.newHashSet();
+    sourceCols.addAll(keySpec.fieldNames());
+    // partitionSpec.fields().stream()
+    //     .map(f -> schema.findField(f.sourceId()))
+    //     .map(Types.NestedField::name)
+    //     .forEach(sourceCols::add);
+
+    for (GenericRecord r : upsertSource) {
+      GenericRecord t = it.next();
+      sourceCols.forEach(col -> t.setField(col, r.getField(col)));
+    }
+    return source;
+  }
+}

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/TestUpsert.java
@@ -62,8 +62,8 @@ public class TestUpsert extends SparkTestBase {
   private static final PartitionSpec partitionSpec = PartitionSpec.builderFor(schema)
       .identity("pt").build();
 
-  private static final int newRecordSize = 300;
-  private static final int upsertRecordSize = 200;
+  private static final int newRecordSize = 30;
+  private static final int upsertRecordSize = 20;
 
   @Parameterized.Parameters
   public static List<Object[]> arguments() {
@@ -92,9 +92,9 @@ public class TestUpsert extends SparkTestBase {
               .withSequencePrimaryKey(primaryKeySpec)
               .withRandomDate("pt")
               .build();
-          List<GenericRecord> target = generator.records(1000);
+          List<GenericRecord> target = generator.records(100);
           List<GenericRecord> source = upsertSource(
-              target, generator, primaryKeySpec, partitionSpec, newRecordSize, upsertRecordSize);
+              target, generator, primaryKeySpec, partitionSpec, upsertRecordSize, newRecordSize);
 
           return new Object[] {target, source, primaryKeySpec, duplicateCheck};
         })
@@ -125,8 +125,9 @@ public class TestUpsert extends SparkTestBase {
 
     for (GenericRecord r : upsertSource) {
       GenericRecord t = it.next();
-      sourceCols.forEach(col -> t.setField(col, r.getField(col)));
+      sourceCols.forEach(col -> r.setField(col, t.getField(col)));
     }
+    source.addAll(upsertSource);
     return source;
   }
 
@@ -174,6 +175,7 @@ public class TestUpsert extends SparkTestBase {
 
   @After
   public void after() {
+    sql("USE spark_catalog");
     sql("DROP VIEW IF EXISTS " + initView);
     sql("DROP VIEW IF EXISTS " + sourceView);
     ArcticCatalog catalog = catalog(catalogNameArctic);
@@ -188,7 +190,7 @@ public class TestUpsert extends SparkTestBase {
     sql("INSERT OVERWRITE " + database + "." + table + " SELECT * FROM " + initView);
     sql("insert into table " + database + '.' + table + " SELECT * FROM " + sourceView);
 
-    rows = sql("SELECT * FROM " + database + "." + table);
+    rows = sql("SELECT * FROM " + database + "." + table + " ORDER BY order_key ");
     Assert.assertEquals(initialize.size() + newRecordSize, rows.size());
   }
 }

--- a/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/util/RecordGenerator.java
+++ b/spark/v3.1/spark/src/test/java/com/netease/arctic/spark/util/RecordGenerator.java
@@ -33,9 +33,12 @@ import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.time.temporal.ChronoUnit;
 import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.Random;
 import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
 public class RecordGenerator {
 
@@ -66,6 +69,12 @@ public class RecordGenerator {
     return record;
   }
 
+  public List<GenericRecord> records(int size) {
+    return IntStream.range(0, size).boxed()
+        .map(x -> this.newRecord())
+        .collect(Collectors.toList());
+  }
+
   public static Builder buildFor(Schema schema) {
     return new Builder(schema);
   }
@@ -73,6 +82,8 @@ public class RecordGenerator {
   public static class Builder {
     final Schema schema;
     final Map<String, ValueGenerator> valueGeneratorMap = Maps.newHashMap();
+
+    long seed = 0;
 
     protected Builder(Schema schema) {
       this.schema = schema;
@@ -90,8 +101,18 @@ public class RecordGenerator {
       return this;
     }
 
+    public Builder withRandomDate(String cols) {
+      valueGeneratorMap.put(cols, new RandomDateString(this.seed));
+      return this;
+    }
+
+    public Builder withSeed(long seed) {
+      this.seed = seed;
+      return this;
+    }
+
     public RecordGenerator build() {
-      return new RecordGenerator(schema, valueGeneratorMap, 0);
+      return new RecordGenerator(schema, valueGeneratorMap, this.seed);
     }
   }
 
@@ -111,7 +132,7 @@ public class RecordGenerator {
     @Override
     public Object get(Type type) {
       Object value = RandomUtil.generatePrimitive(type.asPrimitiveType(), random);
-    
+
       switch (type.typeId()) {
         case TIME:
           return LocalTime.ofNanoOfDay((long) value * 1000);
@@ -121,6 +142,8 @@ public class RecordGenerator {
           } else {
             return EPOCH.plus((long) value, ChronoUnit.MICROS).toLocalDateTime();
           }
+        case DATE:
+          return EPOCH.plus((int) value, ChronoUnit.SECONDS).toLocalDate();
       }
       return value;
     }
@@ -143,6 +166,18 @@ public class RecordGenerator {
           return (int) value.incrementAndGet();
       }
       return super.get(type);
+    }
+  }
+
+  static class RandomDateString extends RandomValueGenerator {
+
+    public RandomDateString(long seed) {
+      super(seed);
+    }
+
+    @Override
+    public Object get(Type type) {
+      return super.get(Types.DateType.get()).toString();
     }
   }
 }

